sdks/java/io/rrio/src/main/java/org/apache/beam/io/requestresponse/Throttle.java

```diff
@@ -0,0 +1,469 @@
+    }
+
+    return input
+        // Step 1. Break up the PCollection into fixed channels assigned to an int key [0,
+        // Rate::numElements).
+        .apply(AssignChannelFn.class.getSimpleName(), assignChannels())
+        .apply(groupingStepName, groupingTransform)
+
+        // Step 2. Convert KV<Integer, Iterable<RequestT>> to KV<Integer, List<RequestT>>.
+        // Working with a List<RequestT> is cleaner than an Iterable in Splittable DoFns.
+        // IterableCoder uses a List for IterableLikeCoder's structuralValue.
+        .apply("ConvertToList", toList())
+        .setCoder(kvCoder)
+
+        // Step 3. Apply the splittable DoFn that performs the actual work of throttling.
+        .apply(ThrottleFn.class.getSimpleName(), throttle());
+  }
+
+  private ParDo.SingleOutput<KV<Integer, List<RequestT>>, RequestT> throttle() {
+    return ParDo.of(new ThrottleFn());
+  }
+
+  @DoFn.BoundedPerElement
+  private class ThrottleFn extends DoFn<KV<Integer, List<RequestT>>, RequestT> {
+    private @MonotonicNonNull Counter inputElementsCounter = null;
+    private @MonotonicNonNull Counter outputElementsCounter = null;
+
+    @Setup
+    public void setup() {
+      if (configuration.getCollectMetrics()) {
+        inputElementsCounter = Metrics.counter(Throttle.class, INPUT_ELEMENTS_COUNTER_NAME);
+        outputElementsCounter = Metrics.counter(Throttle.class, OUTPUT_ELEMENTS_COUNTER_NAME);
+      }
+    }
+
+    /**
+     * Instantiates an initial {@link RestrictionTracker.IsBounded#BOUNDED} {@link OffsetRange}
+     * restriction from [-1, {@link List#size()}). Defaults to [-1, 0) for null {@link
+     * KV#getValue()} elements.
+     */
+    @GetInitialRestriction
+    public OffsetRange getInitialRange(@Element KV<Integer, List<RequestT>> element) {
+      int size = 0;
+      if (element.getValue() != null) {
+        size = element.getValue().size();
+      }
+      incIfPresent(inputElementsCounter, size);
+      return new OffsetRange(-1, size);
+    }
+
+    /**
+     * The {@link GetInitialWatermarkEstimatorState} initializes to this DoFn's output watermark to
+     * a negative infinity timestamp via {@link BoundedWindow#TIMESTAMP_MIN_VALUE}. The {@link
+     * Instant} returned by this method provides the runner the value is passes as an argument to
+     * this DoFn's {@link #newWatermarkEstimator}.
+     */
+    @GetInitialWatermarkEstimatorState
+    public Instant getInitialWatermarkState() {
+      return BoundedWindow.TIMESTAMP_MIN_VALUE;
+    }
+
+    /**
+     * This DoFn uses a {@link WatermarkEstimators.Manual} as its {@link NewWatermarkEstimator},
+     * instantiated from an {@link Instant}. The state argument in this method comes from the return
+     * of the {@link #getInitialWatermarkState}.
+     */
+    @NewWatermarkEstimator
+    public WatermarkEstimator<Instant> newWatermarkEstimator(
+        @WatermarkEstimatorState Instant state) {
+      return new WatermarkEstimators.Manual(state);
```

Comment:
One question I have (genuinely not sure of the answer) is how putting a watermark estimator in the middle of a pipeline impacts the previous watermark. For example, lets say I have the following structure:

```
1) Read from Pub/Sub
2) Throttle
3) GroupByKey
```

And I have a constant stream of elements coming from pub/sub, will the GroupByKey _ever_ fire? Basically, I can imagine the pubsub watermark slowly rising, but since Throttle initializes a new estimator each time it encounters an element, its watermark might never advance (until its processed all elements, which would never happen).

The longer I stare at this (have been looking at this PR for a while this afternoon), the more I'm just generally concerned with using an SDF this way. An SDF is by nature meant to process a _single_ element (or at least a bounded set of elements). Here we're using an SDF on a potentially unbounded dataset in an environment where we don't control the windowing, triggering, or watermark, and I basically have no clue how it will behave (but I'd guess poorly).

I think I've expressed my preference for a timers based approach before (and saying this doesn't work in batch), but I think these concerns just reaffirm that preference.